src/wp-includes/class-wp-query.php

```diff
@@ -3281,16 +3284,36 @@ public function get_posts() {
 			$this->posts = array_map( 'get_post', $this->posts );
 		}
 
-		if ( $q['cache_results'] && $id_query_is_cacheable && ! $cache_found ) {
+
+		if ( $q['cache_results'] ) {
 			$post_ids = wp_list_pluck( $this->posts, 'ID' );
 
-			$cache_value = array(
-				'posts'         => $post_ids,
-				'found_posts'   => $this->found_posts,
-				'max_num_pages' => $this->max_num_pages,
-			);
+			if ( $this->posts && $needs_cache_prime ) {
+				$non_cached_ids = _get_non_cached_ids( $post_ids, 'posts' );
+				if ( ! empty( $non_cached_ids ) ) {
+					$fresh_posts = array();
+					foreach ( $this->posts as $_post ) {
+						if ( in_array( $_post->ID, $non_cached_ids, true ) ) {
+							$fresh_posts[] = $_post;
+						}
+					}
+
+					if ( $fresh_posts ) {
+						// Despite the name, update_post_cache() expects an array rather than a single post.
+						update_post_cache( $fresh_posts );
+					}
+				}
```

Comment:
I don't see this as duplicated code, it does something different. I don't like changing the interface of `_prime_post_caches` and it something I am trying to avoid. 

The point here, is the posts need to be primed only some of the time. For one, the post maybe already be loaded from another query.